File: java/library/src/main/java/com/v2soft/styxlib/server/tcp/TCPServerChannelDriver.java
package com.v2soft.styxlib.server.tcp;

import com.v2soft.styxlib.exceptions.StyxException;
import com.v2soft.styxlib.server.ClientDetails;

import java.io.IOException;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.ServerSocket;
import java.net.UnknownHostException;
import java.nio.channels.SelectionKey;
import java.nio.channels.Selector;
import java.nio.channels.ServerSocketChannel;
import java.nio.channels.SocketChannel;
import java.util.Collection;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.Stack;

/**
 * Created by V.Shcryabets on 5/20/14.
 *
 * @author V.Shcryabets (vshcryabets@gmail.com)
 */
public class TCPServerChannelDriver extends TCPChannelDriver {
    protected ServerSocketChannel mChannel;
    protected Stack<SocketChannel> mNewConnetions, mReadable;
    protected Map<SocketChannel, ClientDetails> mClientStatesMap;
    protected int mLastClientId = 1;

    public TCPServerChannelDriver(String address, int port) {
        super(address, port);
        mNewConnetions = new Stack<>();
        mReadable = new Stack<>();
        mClientStatesMap = new HashMap<>();
    }

    @Override
    public void prepareSocket() throws StyxException {
        InetAddress address = null;
        try {
            address = InetAddress.getByName(mAddress);
        } catch (UnknownHostException e) {
            throw new StyxException(StyxException.DRIVER_CANT_RESOLVE_NAME);
        }
        InetSocketAddress socketAddress = new InetSocketAddress(address, mPort);
        try {
            mChannel = ServerSocketChannel.open();
        } catch (IOException e) {
            throw new StyxException(StyxException.DRIVER_CREATE_ERROR);
        }
        ServerSocket socket = mChannel.socket();
        try {
            socket.bind(socketAddress);
        } catch (IOException e) {
            throw new StyxException(StyxException.DRIVER_BIND_ERROR, String.format(
                    "Can't bind socket in TCPServerChannelDriver::prepareSocket() %s",
                    e.getMessage()));
        }
        try {
            socket.setReuseAddress(true);
            socket.setSoTimeout(getTimeout());
            mChannel.configureBlocking(false);
        } catch (IOException e) {
            throw new StyxException(StyxException.DRIVER_CONFIGURE_ERROR);
        }
    }

    @Override
    public boolean isStarted() {
        return isWorking;
    }

    @Override
    public void run() {
        isWorking = true;
        Selector selector = null;
        try {
            selector = Selector.open();
            mChannel.register(selector, SelectionKey.OP_ACCEPT);

            while ( isWorking ) {
                try {
                    if ( !selector.isOpen() ) {
                        break;
                    }
                    selector.select();
                    if ( !selector.isOpen() ) {
                        break;
                    }
                    Iterator<SelectionKey> iterator = selector.selectedKeys().iterator();
                    while ( iterator.hasNext() ) {
                        SelectionKey key = iterator.next();
                        iterator.remove();
                        if ( key.isAcceptable() ) {
                            SocketChannel clientChannel = mChannel.accept();
                            clientChannel.configureBlocking(false);
                            clientChannel.register(selector, SelectionKey.OP_READ);
                            mNewConnetions.push(clientChannel);
                        } else if ( key.isReadable() ) {
                            SocketChannel clientChannel = (SocketChannel) key.channel();
                            mReadable.push(clientChannel);
                        } else if ( key.isValid() ) {
                            System.out.println("Key invalid");
                        }
                    }
                    processEventsQueue();
                } catch (IOException e) {
                    // this is ok
                    e.printStackTrace();
                } catch (Exception e) {
                    e.printStackTrace();
                    break;
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            try {
                if (selector != null) {
                    selector.close();
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
            try {
                closeSocket();
            } catch (StyxException e) {
                e.printStackTrace();
            }
        }
        mChannel = null;
        isWorking = false;
    }

    private void setNonBlocking(SocketChannel channel) throws IOException {
        channel.configureBlocking(false);
    }

    public void closeSocket() throws StyxException {
        if (mChannel == null) {
            return;
        }
        try {
            mChannel.close();
            mChannel = null;
        } catch (IOException e) {
            throw new StyxException(StyxException.DRIVER_CLOSE_ERROR);
        }
    }

    protected void processEventsQueue() throws IOException {
        // new connections
        for (SocketChannel channel : mNewConnetions) {
            setNonBlocking(channel);
            TCPClientDetails client = new TCPClientDetails(channel, this,
                    mIOUnit, mLastClientId++);
            mRMessageHandler.addClient(client);
            mTMessageHandler.addClient(client);
            mClientStatesMap.put(channel, client);
        }
        mNewConnetions.clear();
        // new readables
        for (SocketChannel channel : mReadable) {
            final TCPClientDetails details = (TCPClientDetails) mClientStatesMap.get(channel);
            if ( readSocket(details) ) {
                mTMessageHandler.removeClient(details);
                mRMessageHandler.removeClient(details);
                mClientStatesMap.remove(channel);
                channel.close();
            }
        }
        mReadable.clear();
    }

    @Override
    public Collection<ClientDetails> getClients() {
        return mClientStatesMap.values();
    }

    @Override
    public boolean isConnected() {
        return true;
    }
}

File: native/V2StyxLib/src/server/tcp/TCPServerChannelDriver.cpp
/*
 * TCPServerChannelDriver.cpp
 *
 */
#include "server/tcp/TCPServerChannelDriver.h"
#include "server/tcp/TCPClientDetails.h"
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netdb.h>
#include <fcntl.h>
#include <unistd.h>
#include <strings.h>
#include "utils/Log.h"

TCPServerChannelDriver::TCPServerChannelDriver(StyxString address, uint16_t port) 
    : TCPChannelDriver(address, port), mSocket(INVALID_SOCKET), mLastClientId(1) {
}

TCPServerChannelDriver::~TCPServerChannelDriver() {
}

void TCPServerChannelDriver::prepareSocket() throw(StyxException) {
    int sockfd = ::socket(AF_INET, SOCK_STREAM, 0);
	if (sockfd < 0) {
		  throw StyxException(DRIVER_CREATE_ERROR);
	}   
	struct hostent *server;
	server = ::gethostbyname(mAddress.c_str());
	if (server == NULL) {
		throw StyxException(DRIVER_CANT_RESOLVE_NAME);
	}

    struct sockaddr_in serv_addr;
    bzero((char *) &serv_addr, sizeof(serv_addr));
    serv_addr.sin_family = AF_INET;
    bcopy((char *)server->h_addr,
        (char *)&serv_addr.sin_addr.s_addr, server->h_length);
    serv_addr.sin_port = htons(mPort);
    int result = ::bind(sockfd, (struct sockaddr *) &serv_addr, sizeof(serv_addr));
    if (result < 0) {
#ifdef WIN32
		closesocket(res);
#else
		::close(sockfd);
#endif
        throw StyxException(DRIVER_BIND_ERROR, 
            "Can't bind socket in TCPServerChannelDriver::prepareSocket() %d", errno);
    }
#warning set socket timeout
    mSocket = sockfd;

    int reuse_addr = 1;
	if (::setsockopt(sockfd, SOL_SOCKET, SO_REUSEADDR, (const char*)&reuse_addr, sizeof(reuse_addr)) != 0) {
        throw StyxException(DRIVER_CONFIGURE_ERROR);
    }

    // enable non-blocking mode
    setNonBlocking(sockfd);
}

size_t TCPServerChannelDriver::getMaxPendingQueue() {
    return 5;
}

std::vector<ClientDetails*> TCPServerChannelDriver::getClients() {
    std::vector<ClientDetails*> result;
    for (std::map<Socket, ClientDetails*>::iterator it = mClientStatesMap.begin();
        it != mClientStatesMap.end(); it++) {
            result.push_back(it->second);
        }
    return result;
}

bool TCPServerChannelDriver::isConnected() {
    return true;
}

bool TCPServerChannelDriver::isStarted() {
    return isWorking;
}

StyxString TCPServerChannelDriver::toString() {
    return StyxString("TCPServerChannelDriver");
}

void* TCPServerChannelDriver::run() {
    isWorking = true;
    ::listen(mSocket, getMaxPendingQueue());
    // Since we start with only one socket, the listening socket, it is the highest socket so far.
	Socket highsock = mSocket;
    fd_set socks;
    struct timeval timeout;
    while (isWorking && !mThread->isInterrupted()) {
        FD_ZERO(&socks);
        FD_SET(mSocket, &socks);

        for (std::map<Socket, ClientDetails*>::iterator it = mClientStatesMap.begin(); 
            it != mClientStatesMap.end(); it++ ) {
            if ( it->first > highsock ) {
                highsock = it->first;
            }
            FD_SET(it->first, &socks);
        }

        timeout.tv_sec = 1;
        timeout.tv_usec = 0;
        int readsocks = ::select(highsock + 1, &socks, (fd_set*) 0, (fd_set*) 0, &timeout);

        if (readsocks < 0) {
			// something wrong with our socket
			break;
		} else if (readsocks > 0) {
            if (FD_ISSET(mSocket, &socks)) {
				Socket clientChannel = ::accept(mSocket, NULL, NULL);
				if ( clientChannel > 0 ) {
                    printf("TCPServer got client %d\n", clientChannel);
					setNonBlocking(clientChannel);
#warning we can add new socket to FD_SET here
                    mNewConnetions.push(clientChannel);
				}
			}
            for (std::map<Socket, ClientDetails*>::iterator it = mClientStatesMap.begin(); 
                it != mClientStatesMap.end(); it++ ) {
				if (FD_ISSET(it->first, &socks)) {
                    printf("TCPServer got readable %d\n", it->first);
                    mReadable.push(it->first);
				}
			}
            printf("TCPServer processEventsQueue enter\n");
            processEventsQueue();
        }
    }
    ::close(mSocket);
    mSocket = INVALID_SOCKET;
    isWorking = false;
}

void TCPServerChannelDriver::setNonBlocking(Socket socket) throw(StyxException) {
#ifdef WIN32
	// If iMode!=0, non-blocking mode is enabled.
	u_long mode=1;
	ioctlsocket(socket, FIONBIO, &mode);
#else
	int opts;
	opts = ::fcntl(socket, F_GETFL);
	if (opts < 0) {
		throw StyxException(DRIVER_CONFIGURE_ERROR);
	}
	opts = (opts | O_NONBLOCK);
	if (::fcntl(socket, F_SETFL,opts) < 0) {
		throw StyxException(DRIVER_CONFIGURE_ERROR);
	}
#endif
}

void TCPServerChannelDriver::closeSocket() throw(StyxException) {
    if (mSocket == INVALID_SOCKET) {
        throw StyxException(DRIVER_CLOSE_ERROR);
    }
    if (::close(mSocket) < 0) {
        throw StyxException(DRIVER_CLOSE_ERROR);
    }
    mSocket = INVALID_SOCKET;
}

void TCPServerChannelDriver::processEventsQueue() throw(StyxException) {
    // new connections
    while (!mNewConnetions.empty()) {
        Socket channel = mNewConnetions.front();
        mNewConnetions.pop();
        setNonBlocking(channel);
        TCPClientDetails* client = new TCPClientDetails(channel, this,
                mIOUnit, mLastClientId++);
        mRMessageHandler->addClient(client);
        mTMessageHandler->addClient(client);
        mClientStatesMap[channel] = client;
    }
    // new readables
    while (!mReadable.empty()) {
        Socket channel = mReadable.front();
        mReadable.pop();
        std::map<Socket, ClientDetails*>::iterator it = mClientStatesMap.find(channel);
        if (it != mClientStatesMap.end()) {
            printf("TCPServer find client\n");
            if (readSocket((TCPClientDetails*)it->second)) {
                mTMessageHandler->removeClient(it->second);
                mRMessageHandler->removeClient(it->second);
                mClientStatesMap.erase(channel);
                #warning delete TCPClientDetails
#warning we should move close logic somewhere outside
                ::close(channel);
            }
        }
    }
}